File: hub/sdks/node/src/TickerSdkOptions.ts
export const TICKERQ_SDK_CONSTANTS = {
    HubGrpcBaseUrl: 'https://grpc.hub.tickerq.net/',
    SdkVersion: '1.0.0',
    SdkType: 'nodejs',
} as const;

export interface TickerSdkLogCaptureOptions {
    enabled: boolean;
    minLevel: 'trace' | 'debug' | 'info' | 'warn' | 'error' | 'critical';
}

export class TickerSdkOptions {
    /** Scheduler worker-stream URL — set after sync with Hub. */
    apiUri: string | null = null;

    /** Fixed Hub gRPC URL. */
    readonly hubControlUri: string = TICKERQ_SDK_CONSTANTS.HubGrpcBaseUrl;

    /** HMAC-SHA256 worker-stream signature key — set after Hub sync. */
    webhookSignature: string | null = null;

    /** Single Hub-issued SDK token. */
    apiKey: string | null = null;

    /** Identifier for this application node. */
    nodeName: string = process.env.COMPUTERNAME
        ?? process.env.HOSTNAME
        ?? 'tickerq-node';

    /** gRPC operation timeout in milliseconds (default: 30000). */
    timeoutMs: number = 30_000;

    /** Allow self-signed SSL certificates for local scheduler worker streams. */
    allowSelfSignedCerts: boolean = false;

    /** Per-execution log forwarding settings. */
    logCapture: TickerSdkLogCaptureOptions = {
        enabled: true,
        minLevel: 'info',
    };

    setApiKey(apiKey: string): this {
        this.apiKey = apiKey;
        return this;
    }

    setNodeName(nodeName: string): this {
        if (!nodeName || !nodeName.trim()) {
            throw new Error('TickerQ SDK: NodeName cannot be empty.');
        }
        this.nodeName = nodeName;
        return this;
    }

    setTimeoutMs(timeoutMs: number): this {
        this.timeoutMs = timeoutMs;
        return this;
    }

    setAllowSelfSignedCerts(allow: boolean): this {
        this.allowSelfSignedCerts = allow;
        return this;
    }

    setLogCapture(options: Partial<TickerSdkLogCaptureOptions>): this {
        this.logCapture = { ...this.logCapture, ...options };
        return this;
    }

    validate(): void {
        if (!this.apiKey) {
            throw new Error('TickerQ SDK: ApiKey is required. Call setApiKey() with the Hub-issued tq_sdk_* token.');
        }
    }
}

File: hub/sdks/dotnet/TickerQ.SDK/TickerSdkOptions.cs
using Microsoft.Extensions.Logging;

namespace TickerQ.SDK;

/// <summary>
/// Controls per-execution ILogger capture: every log call made inside a
/// [TickerFunction] body is forwarded to the dashboard so customers can watch
/// their own logs in real time. Capture happens on top of the host's existing
/// log providers — console / AppInsights / etc. continue to receive the same
/// log calls untouched.
/// </summary>
public sealed class TickerSdkLogCaptureOptions
{
    /// <summary>Forward ILogger output from inside a ticker execution to the dashboard. Default true.</summary>
    public bool Enabled { get; set; } = true;

    /// <summary>Minimum log level to forward. Default Information.</summary>
    public LogLevel MinLevel { get; set; } = LogLevel.Information;
}

public class TickerSdkOptions
{
    /// <summary>
    /// Worker-stream target URL — set by the boot-time function-sync to the Scheduler's
    /// ApplicationUrl returned by the Hub. Null until SyncAsync completes.
    /// </summary>
    internal Uri? ApiUri { get; set; }

    /// <summary>
    /// gRPC base URL of the Hub. Used for both the boot-time SyncNodesFunctions RPC
    /// and the persistent SDK control stream. Locked to the production endpoint —
    /// not user-overridable.
    /// </summary>
    internal Uri HubControlUri => new Uri(TickerQSdkConstants.HubGrpcBaseUrl);

    internal string? WebhookSignature { get; set; }
    /// <summary>
    /// Single bearer-style token issued by the Hub (e.g. <c>tq_sdk_AbCdEf...</c>).
    /// Sent as <c>x-api-key</c> metadata on every gRPC call. Hub looks the token up
    /// by SHA-256 hash; the SDK never sees a paired secret.
    /// </summary>
    internal string? ApiKey { get; private set; }
    /// <summary>
    /// Display name for this SDK node, shown in the Hub dashboard. Defaults to
    /// <see cref="Environment.MachineName"/>; override via <see cref="SetNodeName"/>
    /// when running multiple SDK instances per host so they're distinguishable.
    /// </summary>
    internal string NodeName { get; private set; } = Environment.MachineName;
    /// <summary>
    /// SDK runtime identifier sent to the Hub during function-sync; drives the
    /// dashboard's per-node logo. Hardcoded — this is the .NET SDK.
    /// </summary>
    internal const string SdkType = "dotnet";

    /// <summary>
    /// Per-execution log capture settings. Mutable so customers can flip behavior
    /// (e.g. disable forwarding, raise minimum level) before <c>AddTickerQSdk</c>
    /// returns. Default: enabled at Information.
    /// </summary>
    public TickerSdkLogCaptureOptions LogCapture { get; } = new();

    /// <summary>
    /// Sets the single bearer-style API token (e.g. <c>tq_sdk_AbCdEf...</c>) issued by
    /// the Hub. The token alone identifies which environment this SDK belongs to.
    /// </summary>
    public TickerSdkOptions SetApiKey(string apiKey)
    {
        ApiKey = apiKey ?? throw new ArgumentNullException(nameof(apiKey));
        return this;
    }

    public TickerSdkOptions SetNodeName(string nodeName)
    {
        NodeName = string.IsNullOrWhiteSpace(nodeName) ? throw new ArgumentNullException(nameof(nodeName)) : nodeName;
        return this;
    }

    /// <summary>
    /// Validates that all required configuration options are set.
    /// </summary>
    /// <exception cref="InvalidOperationException">Thrown when required options are missing.</exception>
    internal void Validate()
    {
        if (string.IsNullOrWhiteSpace(ApiKey))
            throw new InvalidOperationException(
                "TickerQ SDK configuration is invalid: ApiKey is required. " +
                "Call SetApiKey() with the Hub-issued token (e.g. tq_sdk_...).");
    }
}
